### PR TITLE
Error handling and LogEventListener implementation

### PR DESCRIPTION
**Pull request by prashantkedia**

### CODE REVIEW

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt [hunk 1]
```diff
@@ -355,21 +385,21 @@ class VerloopFragment : Fragment() {
     @JavascriptInterface
     @Throws(JSONException::class)
     fun onButtonClick(json: String) {
-        Log.d(TAG, " onButtonClick $json")
+        logEvent(LogLevel.INFO, " onButtonClick $json")
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt]: These should be debug events

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt [hunk 2]
```diff
@@ -314,8 +346,8 @@ class VerloopFragment : Fragment() {
     }
 
     private fun callJavaScript(script: String) {
-        Log.d(TAG, "javascript: $script")
-        mWebView?.evaluateJavascript(script, null)
+        logEvent(LogLevel.DEBUG, "Calling javascript: $script")
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt]: Make a list of constant strings which will be sent as a payload. If there are variables which you are passing, keep it as a json in the third parameter. Eg: [LogLevel.DEBUG, JS_CALL, {script: some_script}]

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt [hunk 3]
```diff
@@ -275,11 +275,43 @@ class VerloopFragment : Fragment() {
         }
         val uri = uriBuilder.build()
         Log.d(TAG, uri.toString())
-        mWebView?.loadUrl(uri.toString())
+        onLoadStart()
+        mWebView.loadUrl(uri.toString())
+    }
+
+    private fun onLoadStart() {
+        loading = true
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt:286]: Put more of debug events, eg: here

**Review by raghav39 — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt [hunk 4]
```diff
@@ -106,90 +123,97 @@ class VerloopFragment : Fragment() {
...
-                progressBar?.visibility = View.VISIBLE
-                super.onPageStarted(view, url, favicon)
-            }
-
-            override fun onPageFinished(view: WebView?, url: String?) {
-                Log.d(TAG, "onPageFinished")
-                progressBar?.visibility = View.GONE
-                super.onPageFinished(view, url)
+                error?.let {
+                    logEvent(LogLevel.ERROR, it.toString())
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt]: Why are we not passing the const string as second param?

File: sdk/src/main/java/io/verloop/sdk/Verloop.kt [hunk 5]
```diff
@@ -118,7 +112,6 @@ class Verloop(val context: Context, var verloopConfig: VerloopConfig) {
             private const val TAG = "VerloopInterface"
         }
 
-        @JavascriptInterface
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/Verloop.kt:121]: Don't we need these annotations anymore?

**Review by raghav39 — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/verloop/sdk/Verloop.kt (hunk 5 above)
Comment (prashantkedia) [reply sdk/src/main/java/io/verloop/sdk/Verloop.kt:121]: No, it is not required.

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopActivity.kt [hunk 6]
```diff
@@ -82,14 +82,12 @@ class VerloopActivity : AppCompatActivity() {
     }
 
     override fun onResume() {
-        Log.d(TAG, "onResume")
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopActivity.kt:85]: There are other Log.d in this activity, can we put all of them in to our new logListener for debug flag?

File: sdk/src/main/java/io/verloop/sdk/viewmodel/MainViewModel.kt [hunk 7]
```diff
@@ -26,4 +27,10 @@ class MainViewModel(var configKey: String?, var repository: VerloopRepository) :
             Verloop.eventListeners[configKey]?.onURLClick(json)
         }
     }
+
+    fun logEvent(event: LogEvent) {
+        configKey.let {
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/viewmodel/MainViewModel.kt:39]: Use Log.d here along with callback

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt [hunk 8]
```diff
@@ -35,20 +39,33 @@ import java.util.*
...
+    private var configKey: String? = null
+    private var config: VerloopConfig? = null
+    private var loading: Boolean = false
 
     companion object {
         private const val TAG = "VerloopFragment"
         private const val ICE_CREAM = 12421
         private const val LOLLIPOP = 12422
 
+        private const val JS_CALL_SET_USER_PARAMS = "SetUserParams"
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt]: Keep these into another file

File: sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt [hunk 9]
```diff
@@ -35,20 +39,33 @@ import java.util.*
...
+        private const val JS_CALL_SET_USER_ID = "SetUserId"
+        private const val JS_CALL_SET_DEPARTMENT = "SetDepartment"
+        private const val JS_CALL_SET_RECIPE = "SetRecipe"
+        private const val JS_CALL_SET_CUSTOM_FIELD = "SetCustomField"
+        private const val JS_CALL_SET_WIDGIT_OPENED = "WidgetOpened"
+        private const val JS_CALL_SET_CLOSE = "CLOSE"
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/ui/VerloopFragment.kt]: Use same case format for all the const

**Review by raghav39 — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/verloop/sdk/Constants.kt [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package io.verloop.sdk
+
+object Constants {
+    internal const val JS_CALL_SET_USER_PARAMS = "JS_CALL: SetUserParams"
+    internal const val JS_CALL_SET_USER_ID = "JS_CALL: SetUserId"
+    internal const val JS_CALL_SET_DEPARTMENT = "JS_CALL: SetDepartment"
+    internal const val JS_CALL_SET_RECIPE = "JS_CALL: SetRecipe"
+    internal const val JS_CALL_SET_CUSTOM_FIELD = "JS_CALL: SetCustomField"
+    internal const val JS_CALL_SET_WIDGET_OPENED = " JS_CALL:WidgetOpened"
```
Comment (raghav39) [inline sdk/src/main/java/io/verloop/sdk/Constants.kt]: Sometimes there is a space after the `:` and other time we are missing it. Let's keep it consistent

**Review by raghav39 — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/verloop/sdk/Constants.kt (hunk 10 above)
Comment (prashantkedia) [reply sdk/src/main/java/io/verloop/sdk/Constants.kt]: Fixed

**Pull request merged**